### PR TITLE
Improve width argument handling in map methods

### PR DESCRIPTION
**Pull request by registerrier**
This provides a solution for cutout order. See issue https://github.com/gammapy/gammapy/issues/1608#issuecomment-411398710

### CODE REVIEW

**Comment by cdeil** [comment]
This way of processing `width` will mean that the width option in cutout behaves inconsistently. In Map.create, so far tuples are usually used, and here passing a tuple of float or Angle objects will lead to a UnitsError. I think we should go through a little helper function and use it consistently where a `width` is taken. See my comments in https://github.com/gammapy/gammapy/issues/1608#issuecomment-409824790

@adonath - I think you're working on gammapy.maps today? Are you preparing a PR for this?

**Comment by registerrier** [comment]
Indeed it works with a list of `Angle`  and breaks with a tuple. Then this needs some type testing.

I also added a test to check for the dimension.

**Comment by registerrier** [comment]
This version runs for an `Angle`, a list or a tuple or `Angle`. 
Of course there is a test, but I suppose it does what we expect. We might also limit to the 2 first entries to avoid issues with longer entries. Or simply fail if `width.size>2`

File: gammapy/maps/wcsnd.py [hunk 1]
```diff
@@ -648,10 +651,19 @@ def cutout(self, position, width, mode='trim', copy=True):
...
         cutout : `~gammapy.maps.WcsNDMap`
             The cutout map itself
-        cutout_slices : Tuple of slice objects (with dimension 1 less than that of the non-spatial axes of the map)
         """
         idx = (0,) * len(self.geom.axes)
 
+        width = _check_width(width)
+
+        # We revert the order to comply with astropy.cutout2D ordering
+        if len(width) == 2:
```
Comment (cdeil) [inline gammapy/maps/wcsnd.py]: There should be no if / else needed here, `_check_width` should always return a length 2 tuple.

So remove this and below just pass this to cutout2d
```
size = width[::-1] * u.deg
```

File: gammapy/maps/wcsnd.py [hunk 2]
```diff
@@ -648,10 +651,19 @@ def cutout(self, position, width, mode='trim', copy=True):
...
+            width = width[::-1]
+        else:
+            raise ValueError("WcsNDMap.cutout: to many entries in width argument")
+
+        width = width*u.deg
+        print(width)
```
Comment (cdeil) [inline gammapy/maps/wcsnd.py]: remove print

File: gammapy/maps/wcsnd.py [hunk 3]
```diff
@@ -635,9 +637,10 @@ def cutout(self, position, width, mode='trim', copy=True):
         ----------
         position : `~astropy.coordinates.SkyCoord`
             Center position of the cutout region.
-        width : tuple of `~astropy.coordinates.Angle`
-            Angular sizes of the region in (lon, lat). If only one value is passed,
-            a square region is extracted. For more options see also `~astropy.nddata.utils.Cutout2D`.
+        width : tuple or list of `~astropy.coordinates.Angle`
```
Comment (cdeil) [inline gammapy/maps/wcsnd.py]: Remove mention of "list" here, we just support length 2 tuples and scalars (please don't add more options).
Suggest to remove the link to Cutout2D now, as our options are now different and reading their width will just confuse users. It's now an implementation detail that might even change later.

File: gammapy/maps/wcs.py [hunk 4]
```diff
@@ -21,6 +21,28 @@
...
+    elif isinstance(width,(list,tuple)):
+        if np.isscalar(width[0]):
+            if len(width)==1:
+                return (float(width[0]),float(width[0]))
+            else:
+                return tuple(width)
+        elif isinstance(width[0], (Angle,Quantity)):
+            return tuple([_.to('deg').value for _ in width])
+        else:
+            raise TypeError("Unsupported input for width")
```
Comment (cdeil) [inline gammapy/maps/wcs.py]: If you want to be kind, you show the user what they passed in, i.e. write this:
```
raise TypeError('Unsupported width: {!r}'.format(width))
```

File: gammapy/maps/wcs.py [hunk 5]
```diff
@@ -21,6 +21,28 @@
     'WcsGeom',
 ]
 
+def _check_width(width):
+    """Convert input width as tuple of float to pass to WcsGeom.create"""
+    if isinstance(width, (Angle,Quantity)):
```
Comment (cdeil) [inline gammapy/maps/wcs.py]: Angle is a Quantity sub-class. You can just write `isinstance(width, Quantity)` here with same effect.

File: gammapy/maps/tests/test_wcs.py [hunk 6]
```diff
@@ -221,3 +222,33 @@ def test_region_mask():
 
     mask = geom.region_mask(regions, inside=False)
     assert np.sum(mask) == 8
+
+
+valid_width_inputs = [
+    (10, (10, 10)),
+    (10*u.deg, (10,10)),
+    ((10,), (10, 10)),
+    ([10, 5], (10, 5)),
```
Comment (cdeil) [inline gammapy/maps/tests/test_wcs.py]: Suggest to drop support for Python lists. Many Numpy functions don't for shape arguments. There's drawbacks and no upsides to support more options for users.

File: gammapy/maps/tests/test_wcs.py [hunk 7]
```diff
@@ -221,3 +222,33 @@ def test_region_mask():
...
+    assert g.npix[0][0] == results[0]
+    assert g.npix[1][0] == results[1]
+
+invalid_width_inputs = [
+    '10 deg', ('10 deg', '5 deg'), np.array([10, 5])
+]
+
+@pytest.mark.parametrize(('width'), invalid_width_inputs)
+def test_incorrect_width(width):
+#    print(_check_width(width))
```
Comment (cdeil) [inline gammapy/maps/tests/test_wcs.py]: remove commented-out print.

**Comment by cdeil** [comment]
@registerrier - There's a merge conflict. Please rebase.

File: gammapy/maps/wcsnd.py (hunk 1 above)
Comment (registerrier) [reply gammapy/maps/wcsnd.py]: Done

File: gammapy/maps/wcsnd.py (hunk 2 above)
Comment (registerrier) [reply gammapy/maps/wcsnd.py]: yes sorry

File: gammapy/maps/wcs.py (hunk 4 above)
Comment (registerrier) [reply gammapy/maps/wcs.py]: Done

**Comment by registerrier** [comment]
I simplified the function along the line you suggest. List is no longer supported.

**Comment by cdeil** [comment]
I thought this was merged already. Merging it now.

@registerrier - Thanks!

**Pull request merged**